appvalidator/specs/webapps.py

```diff
@@ -346,6 +412,35 @@ def _path_valid(self, path, can_be_asterisk=False, can_be_absolute=False,
             # If there was an error parsing the URL, return False.
             return False
 
+    def process_role(self, node):
+        if node == 'langpack' and not self.data.get('type') == 'privileged':
+            self.err.error(
+                err_id=("spec", "webapp", "role_langpack_privileged"),
+                error="`langpack` role is only valid for privileged apps.",
+                description=["`langpacks` role is only valid for "
+                            "privileged applications",
+                            self.MORE_INFO])
+
+    def process_languages_target(self, node):
+        role = self.data.get("role", "")
+        if role != "langpack":
+            self.err.error(
+                err_id=("spec", "webapp", "languages_target_langpacks"),
+                error="`languages-target` is only valid for langpacks.",
+                description=["`languages-provided` is only valid for "
```

Comment:
Good catch. I've kept them separate because I expected to have more stuff to do in them, but yeah, you're right, I should make a common helper in the meantime.